### DISCUSSION TITLE
Consider using semantic versioning?

### DISCUSSION THREAD

**Discussion by jenstroeger** [sqlalchemy/sqlalchemy#11391] [Ideas]
Hello!

The other day I noticed that [v2.0.22](https://github.com/sqlalchemy/sqlalchemy/releases/tag/rel_2_0_22) adds the `get_one()` method to sessions ([docs](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.get_one)) which — as per semantic versioning (semver, [specs](https://semver.org/)) — is a “feature” that would bump the minor version component of the version string (so it would bump from v2.0.21 to v2.1.0).

Scrolling through the Changelog, I noticed the careful use of SQLA components (e.g. `[orm]` or `[postgres]`) paired with `[bug]` and `[usecase]` etc. — but those changes aren’t reflected in the version string itself.

In the past, has there been discussion regarding semver, and is there perhaps a plan to move towards semver?

That would make declaring package/app dependencies on SQLA easier and more fine-grained, e.g. I’d be able to automatically take only bug-fix versions: `sqlalchemy ==2.0.*` while I also would avoid new features.

Thanks!
Jens

**Comment by CaselIT** [comment]
Hi,

It was considered and it would be too impractical, meaning there would not be bigfixes only releases.
At the moment the plan is that 2.1 is the next major release

**Reply by jenstroeger** [reply]
Thank you @CaselIT! After some searching I found this [VERSION NUMBERING SCHEME](https://www.sqlalchemy.org/download.html) section in the documentation 👍🏼

**Comment by zzzeek** [comment]
pure semantic versioning doesn't work for us because we need to add small API improvements like the get_one() method all the time and we can't wait for the longer "feature" cycle.   to mitigate, we tend to call these "use cases"  - they're in between "bug" and "feature" - typically some obvious method that was missing, or some database feature or syntax not directly supported.   These can be considered "bugs" in some circles.   We use the term "use case".   if you look at our point releases, you should see little to nothing tagged as a "feature", it should be "bug" and "use case".   

Then, we have very long "feature" cycles because bigger features come with a lot more internal refactoring, sometimes subtle behavioral changes, and most importantly risk.  So our "major" releases are on the order of 12-18 months and we do major migration guides to give heads up to all the things that might be surprising.  that's where you see "features", like new kinds of loader options and things like that.   this level of release is also where we make moves with deprecations, i.e. removing long deprecated APIs and such.

if we used pure semver our releases would always be "x.y.0", it would always be a "y" bump, and we'd have no digit available for the level of change that we did from 1.x to 2.     so our release cadence is very similar to that of cPython.

for your original concern, "That would make declaring package/app dependencies on SQLA easier and more fine-grained, e.g. I’d be able to automatically take only bug-fix versions: sqlalchemy ==2.0.* while I also would avoid new features." - the reason you want to avoid new features is because of the risk they bring.  The "features" that we put in point releases are called "use case", and they are extremely carefully vetted to ensure they are not introducing new risk.

**Reply by jenstroeger** [reply]
I appreciate you elaborating @zzzeek, thank you! I’m currently declaring the SQLA dependency as `"sqlalchemy >=2.0.26,<2.1.0"` which should be a fairly safe range considering your cadence and versioning scheme…
